fasteners/threads.scad

```diff
@@ -44,32 +59,140 @@ use <scad-utils/transformations.scad>
 
 
 // ----------------------------------------------------------------------------
-// internal - true = clearances for internal thread (e.g., a nut).
-//            false = clearances for external thread (e.g., a bolt).
-//            (Internal threads should be "cut out" from a solid using
-//            difference()).
-// n_starts - Number of thread starts (e.g., DNA, a "double helix," has
-//            n_starts=2).  See wikipedia Screw_thread.
+// Metric Threads in accordance with ISO 68-1
+// ----------------------------------------------------------------------------
+// diameter  - nominal diameter of thread in mm. (e.g. 10 for an M10 thread)
+// pitch     - distance in mm from the crest of one thread to the next.
+// length    - overall axial length of thread in mm.
+// internal  - true = clearances for internal thread (e.g., a nut).
+//             false = clearances for external thread (e.g., a bolt).
+//             (Internal threads should be "cut out" from a solid using
+//             difference()).
+// clearance - additional clearance to add in mm. Teeth are effectively
+//             shifted inwards (outwards) for external (internal) threads by
+//             this amount.
+// n_starts  - Number of thread starts (e.g., DNA, a "double helix," has
+//             n_starts=2).  See Wikipedia Screw_thread.
 module metric_thread (
     diameter = 8,
     pitch = 1,
     length = 1,
     internal = false,
+    clearance = 0,
     n_starts = 1
 )
 {
+    // Tolerancing as per ISO 286
+    // See https://en.wikipedia.org/wiki/IT_Grade
+    function IT_grade (diameter, ITG) =
+        assert(ITG>=1, "ITG has to be between 1 and 16")
+        assert(ITG<=16, "ITG has to be between 1 and 16")
+        let (i = 0.45 * pow(diameter, 1/3) + 0.001 * diameter)
+            pow(10, 0.2*(ITG-1)) * i / 1000;
+
+    // Fundamental deviations as per ISO 286. Formulas obtained from:
+    // https://www.cobanengineering.com/Tolerances/ToleranceBand.asp
+    function EI_es_g (diameter) = 2.5 * pow(diameter, 0.34) / 1000;
+
+    theta = 60;                // V-shape angle, 60º for metric threads
+    H = pitch/2/tan(theta/2);  // height of the V-shape
+    tolerance = internal
+        ? IT_grade(diameter, 6)/2 + 0                  // H6
+        : IT_grade(diameter, 6)/2 + EI_es_g(diameter); // g6
+    D_maj = internal
```

Comment:
Looks good, but it looks like we're unconditionally applying `tolerance` right now. Could we perhaps add a flag to opt-out of applying `tolerance`?

Reply:
I see your point, and you are right that the code silently forces that minimum clearance. Please allow me to explain. A size for size male and female threads will never work, no matter how precise your printer/CNC is. They won't fit. One needs a minimum clearance fit, which is why I forced an H6/g6 fit in this case (which in turn is, quite simplistically, the fit recommended by the standard). The idea behind always applying `tolerance` was two fold:
- That if a novice user was to blindly create two threads, print them and plug them into each other, there's a chance they would actually fit, without the user needing to know anything about fits and clearances.
- That one could buy an off-the-shelf M20 screw, design and print a part with a female M20 thread, again without specifying any fits/clearances, and, if the printer is precise enough, they will fit. So briefly, the idea was to align with the standard to ensure compatibility with other manufactured parts.

Could I suggest, instead of adding just another opt-out parameter, making the `clearance` parameter equal to `undef` by default? Code logic will then be as per below. A user willing no clearance at all will then set `clearance = 0`.
```pseudocode
if clearance == undef:
    Apply an approximate H6/g6 fit clearance
if clearance != undef:
    clearance = 0 will make size for size threads
    clearance = x will make the clearance exactly x, with no hidden extra tolerancing
```

Reply:
Sounds good to me.

I understood your rationale for introducing the clearance by default, but at the same time, I thought there might be situations where the user of the module wants to directly control the amount of slop the thread has. I think this is nicely handled by the default-undef value for `clearance` now.